### PR TITLE
Escape more strings: formspecs, item descriptions, infotexts...

### PR DESCRIPTION
**Pull request by Ekdohibs**
Thus \x1b can be used in the future for translation, for example.

### CODE REVIEW

**Comment by Ekdohibs** [comment]
I'd like this to be merged in 0.4.14 as it is compatibility for future code; marking as milestone.

File: src/guiFormSpecMenu.h [hunk 1]
```diff
@@ -235,10 +236,11 @@ class GUIFormSpecMenu : public GUIModalMenu
 		}
 		TooltipSpec(std::string a_tooltip, irr::video::SColor a_bgcolor,
 				irr::video::SColor a_color):
-			tooltip(a_tooltip),
 			bgcolor(a_bgcolor),
 			color(a_color)
 		{
+			tooltip = unescape_string(wide_to_utf8(
```
Comment (nerzhul) [inline src/guiFormSpecMenu.h]: maybe having a helper for this ?

**Comment by paramat** [comment]
I'd like to help out by reviewing but i don't understand the code, it's best others review.

File: src/util/string.h [hunk 2]
```diff
@@ -387,12 +387,12 @@ inline void str_replace(std::string &str, const std::string &pattern,
...
 
 /**
- * Remove all chat escape sequences in \p s.
+ * Remove all escape sequences in \p s.
  *
  * @param s The string in which to remove escape sequences.
  * @return \p s, with escape sequences removed.
  */
-std::wstring removeChatEscapes(const std::wstring &s);
+std::wstring removeEscapes(const std::wstring &s);
```
Comment (kwolekr) [inline src/util/string.h]: This function should be lower_case_underscore style since it's a standalone function.

File: src/util/string.cpp [hunk 3]
```diff
@@ -729,11 +729,11 @@ static bool parseNamedColorString(const std::string &value, video::SColor &color
 	return true;
 }
 
-std::wstring removeChatEscapes(const std::wstring &s) {
+std::wstring removeEscapes(const std::wstring &s) {
 	std::wstring output;
 	size_t i = 0;
 	while (i < s.length()) {
```
Comment (kwolekr) [inline src/util/string.cpp]: I feel like this function could be structured better.  Also, this deserves a unit test.

File: src/guiFormSpecMenu.h [hunk 4]
```diff
@@ -199,10 +200,10 @@ class GUIFormSpecMenu : public GUIModalMenu
 		FieldSpec(const std::string &name, const std::wstring &label,
 				const std::wstring &fdeflt, int id) :
```
Comment (kwolekr) [inline src/guiFormSpecMenu.h]: fdeflt?  WTF?

File: src/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -1604,14 +1593,14 @@ void GUIFormSpecMenu::parseTooltip(parserData* data, std::string element)
 	std::vector<std::string> parts = split(element,';');
```
Comment (kwolekr) [inline src/guiFormSpecMenu.cpp]: What is the difference between this function `split` and `str_split`?

File: src/guiFormSpecMenu.cpp [hunk 6]
```diff
@@ -1604,14 +1593,14 @@ void GUIFormSpecMenu::parseTooltip(parserData* data, std::string element)
...
 	if (parts.size() == 2) {
 		std::string name = parts[0];
-		m_tooltips[name] = TooltipSpec(unescape_string(parts[1]),
+		m_tooltips[name] = TooltipSpec(parts[1],
```
Comment (kwolekr) [inline src/guiFormSpecMenu.cpp]: Why were the unescape_string() calls removed here?

File: src/util/string.cpp [hunk 7]
```diff
@@ -729,11 +729,11 @@ static bool parseNamedColorString(const std::string &value, video::SColor &color
...
-		if (s[i] == L'\v') {
+		if (s[i] == L'\x1b') {
```
Comment (kwolekr) [inline src/util/string.cpp]: \v is 0x0B while you're changing the new escape character to 0x1B.  Won't this break reverse compatibility?

Comment (Ekdohibs) [inline src/util/string.cpp]: It won't, because these escapes were added but a month ago and were never used for anything yet.

File: src/guiFormSpecMenu.cpp [hunk 8]
```diff
@@ -746,7 +743,8 @@ void GUIFormSpecMenu::parseTable(parserData* data,std::string element)
 		spec.ftype = f_Table;
 
 		for (unsigned int i = 0; i < items.size(); ++i) {
-			items[i] = unescape_string(items[i]);
+			items[i] = unescape_string(wide_to_utf8(
+				remove_escapes(utf8_to_wide(items[i]))));
```
Comment (kwolekr) [inline src/guiFormSpecMenu.cpp]: Doesn't this imply the need for a variant of remove_escapes() that works on a utf-8 string?

File: src/guiFormSpecMenu.cpp (hunk 6 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: The unescape_string calls are now part of the constructor of TooltipSpec.

File: src/guiFormSpecMenu.cpp (hunk 5 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: ```split``` checks for escaped characters before breaking, while ```str_split``` does not.

File: src/unittest/test_utilities.cpp [hunk 9]
```diff
@@ -253,6 +255,13 @@ void TestUtilities::testUTF8()
 		== "the shovel dug a crumbly node!");
 }
 
+void TestUtilities::testRemoveEscapes()
+{
+	UASSERT(remove_escapes(L"abc\x1bXdef") == L"abcdef");
+	UASSERT(remove_escapes(L"abc\x1b(escaped)def") == L"abcdef");
+	UASSERT(remove_escapes(L"abc\x1b((escaped with parenthesis\\))def") == L"abcdef");
```
Comment (kwolekr) [inline src/unittest/test_utilities.cpp]: This tests functionality well, but could you also add some edge cases so I have confidence it'll work?
For example, an incomplete escape sequence at the end of the string, an escape sequence with an unclosed set of parentheses, an escape sequence without a set of parentheses at all, etc.  The point of a test is to try your hardest to break the code, not feed it values you already know work well.

File: src/guiFormSpecMenu.h (hunk 4 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.h]: That was the name of the parameter, I guess it is to avoid a clash with field ```fdefault```.

File: src/unittest/test_utilities.cpp (hunk 9 above)
Comment (Ekdohibs) [inline src/unittest/test_utilities.cpp]: Without parentheses is already the first test, I'm adding the two others.

File: src/guiFormSpecMenu.cpp (hunk 8 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: We might want to add data about a number of characters in escapes (let's say, remove the previous ```n``` characters), and that needs to be consistent so the text has to either be always utf8 or always wide, but not sometimes one and sometimes the other.

File: src/guiFormSpecMenu.h (hunk 4 above)
Comment (kwolekr) [inline src/guiFormSpecMenu.h]: Maybe change it to `default` so it's actually legible?  It'll also be consistent with the rest of the parameters which are otherwise identical to the members, but lacking the f prefix.

File: src/unittest/test_utilities.cpp [hunk 10]
```diff
@@ -253,6 +255,15 @@ void TestUtilities::testUTF8()
...
+	UASSERT(remove_escapes(L"abc\x1b(incomplete") == L"abc");
+	UASSERT(remove_escapes(L"escape at the end\x1b") == L"escape at the end");
+}
```
Comment (kwolekr) [inline src/unittest/test_utilities.cpp]: How about a unit test that has a nested escape sequence, like \x1b(blah\x1b(asdf)blah)dffgd?

File: src/guiFormSpecMenu.cpp (hunk 8 above)
Comment (est31) [inline src/guiFormSpecMenu.cpp]: Wide characters are not uniform across platforms. On some they are wider, on some they are smaller.

File: src/guiFormSpecMenu.h [hunk 11]
```diff
@@ -191,18 +192,26 @@ class GUIFormSpecMenu : public GUIModalMenu
...
+	 *    Irrlicht and we can't unescape it later.
+	 */
 	struct FieldSpec
 	{
 		FieldSpec()
 		{
 		}
 		FieldSpec(const std::string &name, const std::wstring &label,
-				const std::wstring &fdeflt, int id) :
+				const std::wstring &default, int id) :
```
Comment (est31) [inline src/guiFormSpecMenu.h]: :-1: to the rename, this does bad things with syntax coloring.

File: src/guiFormSpecMenu.h [hunk 12]
```diff
@@ -252,18 +261,18 @@ class GUIFormSpecMenu : public GUIModalMenu
 		}
 		StaticTextSpec(const std::wstring &a_text,
 				const core::rect<s32> &a_rect):
-			text(a_text),
 			rect(a_rect),
 			parent_button(NULL)
 		{
+			text = unescape_string(remove_escapes(a_text));
```
Comment (est31) [inline src/guiFormSpecMenu.h]: We have two escape systems now, have you thought about giving it a better name that makes it easier to distinguish the two systems?

File: src/guiFormSpecMenu.h (hunk 11 above)
Comment (kwolekr) [inline src/guiFormSpecMenu.h]: Whoops, that's why it's fdeflt I guess.  Maybe make it default_text instead?

File: src/guiFormSpecMenu.cpp (hunk 8 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: @est31 wait, some characters can span several wide chars?

File: src/guiFormSpecMenu.h (hunk 12 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.h]: What about ```remove_enriched_text_escapes```? Maybe a bit too long though.

File: src/guiFormSpecMenu.cpp (hunk 8 above)
Comment (est31) [inline src/guiFormSpecMenu.cpp]: Yes, on windows there are [surrogate pairs](https://en.wikipedia.org/wiki/Universal_Character_Set_characters#Surrogates). I'm not sure whether linux has them too, or uses the code points.

**Comment by C1ffisme** [comment]
Will this remove this hack?

screenshot_20160328_110840
```
minetest.register_craftitem("tests:newline",{
	description = "Line1 \nLine2 \nLine3 \nLine4",
	inventory_image = "default_wood.png",
})
```

Just thought I'd ask.

**Comment by Ekdohibs** [comment]
@C1ffisme This is not a hack IMHO... Multi-line item descriptions may be useful.

**Comment by C1ffisme** [comment]
@Ekdohibs Yeah, especially for more detailed 'descriptions' or lore:

screenshot_20160416_100131

**Comment by paramat** [comment]
@Ekdohibs removed from milestones, freeze is probably at weekend and i don't think we'll wait for this, however if you're able to get it done in time that's good.

File: src/unittest/test_random.cpp [hunk 13]
```diff
@@ -50,7 +50,7 @@ void TestRandom::runTests(IGameDef *gamedef)
 	TEST(testPseudoRandom);
 	TEST(testPseudoRandomRange);
 	TEST(testPcgRandom);
-	TEST(testPcgRandomRange);
+	//TEST(testPcgRandomRange);
```
Comment (kwolekr) [inline src/unittest/test_random.cpp]: why'd you do this??

File: src/util/string.h [hunk 14]
```diff
@@ -492,6 +484,39 @@ inline std::basic_string<T> unescape_string(std::basic_string<T> &s)
...
 }
 
+/**
+ * Remove all escape sequences in \p s.
+ *
+ * @param s The string in which to remove escape sequences.
+ * @return \p s, with escape sequences removed.
+ */
+template <typename T>
+std::basic_string<T> remove_enriched_text_escapes(const std::basic_string<T> &s) {
```
Comment (kwolekr) [inline src/util/string.h]: Yeah this is way too long of a name... I can't really offer a suggestion for an alternative, but this is too long.  Also, the line is way over 80 characters.  The { should be on the next line too.

File: src/unittest/test_utilities.cpp [hunk 15]
```diff
@@ -253,6 +255,23 @@ void TestUtilities::testUTF8()
 		== "the shovel dug a crumbly node!");
 }
 
+void TestUtilities::testRemoveEscapes()
+{
+	UASSERT(remove_enriched_text_escapes<wchar_t>(
```
Comment (kwolekr) [inline src/unittest/test_utilities.cpp]: Why do you explicitly have the template parameter here but not anywhere else??

**Comment by kwolekr** [comment]
Wait, so does this commit break newlines in tooltips or not?

File: src/unittest/test_random.cpp (hunk 13 above)
Comment (Ekdohibs) [inline src/unittest/test_random.cpp]: Oops, I didn't meant to commit that change. The problem is that that line crashes unit tests.

File: src/util/string.h (hunk 14 above)
Comment (Ekdohibs) [inline src/util/string.h]: ```unescape_enriched```?

File: src/unittest/test_utilities.cpp (hunk 15 above)
Comment (Ekdohibs) [inline src/unittest/test_utilities.cpp]: I think it is because the arguments are constant strings, which are interpreted by the compiler as ```const wchar_t *``` and not ```str::wstring```

**Comment by Ekdohibs** [comment]
This commit changes nothing from the previous behaviour about breaking newlines in tooltips.

File: src/util/string.h (hunk 14 above)
Comment (kwolekr) [inline src/util/string.h]: Sure, that works I guess.

File: src/unittest/test_random.cpp (hunk 13 above)
Comment (kwolekr) [inline src/unittest/test_random.cpp]: Interesting, if you have a backtrace I'd like to have a look at that.

File: src/guiFormSpecMenu.cpp (hunk 5 above)
Comment (ShadowNinja) [inline src/guiFormSpecMenu.cpp]: Should be `str_split_escaped` then.

**Comment by ShadowNinja** [comment]
This should use a proper escape character like `\x1B`.
`\v` has a text formatting function just like, eg, `\n` (although `\v` is probably not properly handled at the moment).

**Comment by Ekdohibs** [comment]
@ShadowNinja it already uses ```\x1b``` instead of the original ```\v```

**Comment by Zeno-** [comment]
:+1:

File: src/guiFormSpecMenu.cpp (hunk 5 above)
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: Probably, however I did not touch this code so I will not change it in that PR; I could however do another one for that.

File: src/guiFormSpecMenu.cpp [hunk 16]
```diff
@@ -2508,11 +2508,11 @@ void GUIFormSpecMenu::drawMenu()
 		if (id != -1 && delta >= m_tooltip_show_delay) {
 			for(std::vector<FieldSpec>::iterator iter =  m_fields.begin();
 					iter != m_fields.end(); ++iter) {
-				if ( (iter->fid == id) && (m_tooltips[iter->fname].tooltip != "") ){
+				if ( (iter->fid == id) && (m_tooltips[iter->fname].tooltip != L"") ){
```
Comment (paramat) [inline src/guiFormSpecMenu.cpp]: Inner brackets not needed? Needs space before curly bracket.

**Comment by paramat** [comment]
Added a line comment.
I think hmmmm should look at this as he had some issues.

**Comment by kwolekr** [comment]
:+1:

**Comment by paramat** [comment]
Code style looks ok now.

**Comment by Zeno-** [comment]
48939df
